### ISSUE TITLE
Stutters of 60 ms when a keyboard input is made after a prolonged game session

### ISSUE THREAD

**Issue by troy0h** [ValveSoftware/gamescope#1292]
![image](https://github.com/ValveSoftware/gamescope/assets/26183563/238fa0d7-3a8b-4881-96c5-77efe0f01fc3)
Any game I run with gamescope, after about an hour, the game will start to have large frametime spikes.
These frametime spikes happen every time a keyboard input is made (both pressing and releasing a key), and are up to 60 ms.
This makes games pretty much unplayable.
I've had this in Team Fortress 2, Lethal Company and Helldivers 2.

**Comment by kisak-valve** [comment]
Hello @troy0h, this reads similarly to https://github.com/ValveSoftware/csgo-osx-linux/issues/2901. Do you happen to have the Steam overlay disabled, and if that's true, does enabling it have an effect?

**Comment by MGThePro** [comment]
I've been experiencing the same issue in cs2. The issue happened only after switching to using gamescope, so I don't think it has anything to do with the game (but I can't 100% rule it out)

I have the steam overlay enabled, but I can't open it so not sure if that could affect it

**Comment by troy0h** [comment]
> Hello @troy0h, this reads similarly to [ValveSoftware/csgo-osx-linux#2901](https://github.com/ValveSoftware/csgo-osx-linux/issues/2901) / [ValveSoftware/Source-1-Games#5767](https://github.com/ValveSoftware/Source-1-Games/issues/5767). Do you happen to have the Steam overlay disabled, and if that's true, does enabling it have an effect?

Hi, I can't even seem to get the steam overlay working in game, I add the `-e` flag, but it doesn't make shift - tab work in games
My lethal company launch options are `gamescope -f -w 2560 -h 1440 -- gamemoderun mangohud %command%`
My helldivers 2 launch options are `gamescope -f -r 165 -w 2560 -h 1440 -e -- gamemoderun mangohud %command% --use-d3d11`
both have the same issue, even though only Helldivers 2 actually has `-e` to enable Steam integration, while Lethal Company doesn't

**Comment by sharkautarch** [comment]
If you don’t have steam overlay disabled in steam’s options for that game, it could still have an effect on gamescope’s performance, even if you are unable to bring up the overlay w/ shift + tab. This is because steam overlay is a dynamic library loaded w/ `LD_PRELOAD`. Due to how `LD_PRELOAD` and environment variables work, this may mean that steam overlay may try to load itself into gamescope…

**Comment by sharkautarch** [comment]
Tho it is definitely possible there’s some weird issue happening in gamescope itself.
I planning on hooking up Tracy profiler into gamescope, which would hopefully reveal the exact source of these sudden slowdowns. Especially since I’ve seen other unique & specific situations where gamescope will stutter for unclear reasons

**Comment by troy0h** [comment]
> If you don’t have steam overlay disabled in steam’s options for that game, it could still have an effect on gamescope’s performance, even if you are unable to bring up the overlay w/ shift + tab. This is because steam overlay is a dynamic library loaded w/ `LD_PRELOAD`. Due to how `LD_PRELOAD` and environment variables work, this may mean that steam overlay may try to load itself into gamescope…
I've experienced it happening with or without the steam overlay, it doesnt make a difference (at least for me)

**Comment by JimBimBum** [comment]
this doesn't seem to be a gamescope or wayland issue, i use x11 and have this issue only with TF2 only with vulkan enabled. seems fine on opengl or i can't tell. only pattern i've found is around 2.5 GBs of ram used tf2 starts doing that. it is clearly has something to do with input but idk what or why.

first joined, no stutters 
![image](https://github.com/ValveSoftware/gamescope/assets/70550505/b6678949-e2a2-4952-aeed-918d1386ee73)
now stutters
![image](https://github.com/ValveSoftware/gamescope/assets/70550505/5e46c60f-6937-4afb-b3d3-5a7c77bc1f60)
![image](https://github.com/ValveSoftware/gamescope/assets/70550505/3d7e3767-033d-44d1-bf47-7efbf0a57838)

**Comment by JimBimBum** [comment]
maybe an sdl3 issue?

**Comment by BigBrotherLovesYou** [comment]
Also suffering from this, both Wayland and X11 sessions.
KDE 6.0.4.1 (so with the fix  referenced [in](https://github.com/ValveSoftware/Source-1-Games/issues/5630)) kernel 6.9.1

**Comment by Joshua-Ashton** [comment]
Please try adding `LD_PRELOAD=""` before launching Gamescope

**Comment by JimBimBum** [comment]
just turning on steam overlay fixes this for me but idk about you though.

**Comment by lsr1006** [comment]
![image](https://github.com/user-attachments/assets/e9a50fc0-a08a-4243-83dd-59df88aae7d7)
I have also experienced this issue, happening on both steam and non-steam games (Arch, Hyprland, 7900XT). Screenshot above is from May (Black Ops 4 - battle.net on Bottles) but I didn't know gamescope was causing the issue until today when it happened again in "Chained Together" on Steam, only when using gamescope. Same scenario as OP - happens after 60-90 ish minutes of play and is for every keyboard input, otherwise is perfectly smooth. I can try using LD_PRELOAD with the launch option next time to see if it helps.

**Comment by sharkautarch** [comment]
@lsr1006 yeah try the LD_PRELOAD="" thing
If it doesn’t fix the issue, I could give you some instructions for profiling gamescope to get insight into what’s causing the problem…

Tho it’d probably be a good idea to first try installing `gamescope-git` from the aur, to double check that the issue persists in the latest commit.

for profiling, gamescope has support for gpuvis, but there’s also another profiler that I’ve been working on hooking up: Tracy profiler. I’m thinking that Tracy profiler may work better for this case…
I’ll give you more info later if you continue to have this issue and you’re up to chasing this weird performance bug…

**Comment by lsr1006** [comment]
@sharkautarch LD_PRELOAD="" does seem to resolve the issue for me, at least in the one game I tried. Issue did not present after 2+ hrs of time in game.

**Comment by JustPlainGarak** [comment]
I put together a workaround that both prevents the frame time issue AND gets the Steam Overlay working.

```LD_PRELOAD="" gamescope --steam -- env LD_PRELOAD="$HOME/.local/share/Steam/ubuntu12_64/gameoverlayrenderer.so" %command%```

All this does is, as the original workaround does, cancel out LD_PRELOAD for gamescope, but then INSTEAD pass the overlay library via LD_PRELOAD to the command inside gamescope. So far, this has worked perfectly in my testing. No unplayable frame time glitches and the Steam Overlay injects and works normally.

- If your game is 32-bit, change ```ubuntu12_64``` to ```ubuntu12_32``` (or just include both paths).
- Pepper in your preferred or required gamescope parameters to taste. Personally I'm using the HDR parameters.

**Comment by junelva** [comment]
For some games, the stuttering that occurs after a period of time can be fixed by using the gamescope option `--backend sdl`. In some cases (especially in wlroots compositors) it may be necessary to use two gamescopes, one with `--backend wayland` and the inner one with `--backend sdl`.

As an example launch option it could be `gamescope --backend wayland -W 1920 -H 1080 -- gamescope --backend sdl -W 1920 -H 1080 -- %command%` or somesuch.

**Comment by Andy-Danderson** [comment]
I have this exact issue, except without gamescope and limited to TF2.

**Comment by junelva** [comment]
I didn't mention this above, but:

You might try disabling audio for the game to see if that is the source of the performance degradation. In my case that was it... and incidentally using the SDL backend *or* disabling audio in the game would help performance.

I believe the launch option for TF2 is `-nosound`.

The intent will not be then to play without sound, but perhaps if that alleviates the issue, there may be something that can be done to wrap or look into why this may be the case. My random guess is that it's an incompatibility with something in Wayland, Pipewire, or perhaps even users who change audio latency permissions on their system (i.e. to use music software or a DAW)...

**Comment by CallieEZ** [comment]
this bug still exists somehow, the LD_PRELOAD workaround is fine but it's really baffling this hasn't been fixed

**Comment by BenjaminBenetti** [comment]
Yup still happening, came hear after getting stuttering just as described after 1 - 2 hours game time, every keyboard input causes a frame time spike. Will try the LD_PRELOAD workaround. 

Edit
--- 
Adding 
```
LD_PRELOAD=""
```
At the start of your steam launch options totally solves the issue (for me at least)

**Comment by Reycko** [comment]
This has started to happen to me after a full system upgrade on Arch Linux, even after a reboot. Both stable and -git

**Comment by jynbug** [comment]
Apologies for not responding with my current workaround for this, consider running steam itself inside of gamescope. There is perhaps nothing particularly secure about this method, but there never was to begin with, with any of these softwares, so there is that at least.

This is the command I use, of course I expect it will not work once this workaround is identified, however for that I was able to continue my work.

`env SDL_VIDEODRIVER=x11 gamescope --backend wayland -r 144 --force-composition -H 720 -F nearest -- steam`

Edit: Notes: If you restrain the resolution as I have, the mouse will most likely not line up with the contents of the window. It's another ancient unfixed bug in gamescope. I have been wanting to fix these bugs for a long time but for a time quit using it. However due to other games changing their own anti-user behavior, I had to change it.

**Comment by riki137** [comment]
Happens on Fedora 44 Silverblue, gamescope rpm package, GNOME 50 Wayland. Periodic stuttering when holding keys (when i hold W, it's lagging at the same times a letter would appear in text editor). AMD GPU.
